keyvault/certrefresher.go

```diff
@@ -0,0 +1,125 @@
+package keyvault
+
+import (
+	"context"
+	//nolint:gosec // sha1 only used to display cert thumbprint in logs for cross-verification with keyvault.
+	"crypto/sha1"
+	"crypto/tls"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type EventualExpirationErr struct {
+	time.Time
+}
+
+func (e *EventualExpirationErr) Error() string {
+	return fmt.Sprintf("could not refresh before expiration on %v", e.Time.String())
+}
+
+type tlsCertFetcher interface {
+	GetLatestTLSCertificate(ctx context.Context, certName string) (tls.Certificate, error)
+}
+
+// CertRefresher offers a mechanism to present the latest version of a tls.Certificate from KeyVault, refreshed at an interval.
+type CertRefresher struct {
+	certName string
+	kvc      tlsCertFetcher
+	logger   *zap.Logger
+	m        sync.RWMutex
+	cert     *tls.Certificate
```

Comment:
It's a small thing, but I like to separate out the mutex and the thing (or things) it's protecting... I usually put the mutex below it.
```suggestion
	logger   *zap.Logger
	
	cert     *tls.Certificate
	m        sync.RWMutex
```